### PR TITLE
feat: type checking in tasks.

### PR DESCRIPTION
**Pull request by peterhuene**
This commit implements full type checking in tasks.

* Detection of cycles in declarations.
* Type calculations of expressions.
* Type checking of string interpolations.

Workflow type checking is forthcoming.

The following have not been implemented yet, which is required for full 1.2 support:

* support `task` name (with "task" hidden type) in commands and output declarations.
* support `input`, `output`, and `hints` hidden types in literal expression evaluation.

Additionally, the following changes are included:

* Refactored `wdl-analysis` to make certain modules public rather than reexporting everything at the root (cleans up the docs a bit).
* Removed DFS space from `DocumentGraph` as it doesn't need to persist after analysis completes.
* Simplified `DocumentScope` to contain only whats needed to express names and their types.
* Removed the restriction on `Map` type representation to allow `Union` as the key type, which allows type checking to work on a literal `{}`, which has calculated type `Map[Union, Union]`.
* Fixed `basename` function to accept a `String` argument.
* Fixed `size` function to accept a `String` argument.
* Removed the `span_of` function in favor of an `AstNode` extension trait.
* Changed the `call_expr` grammar rule to only accept an identifier for the target and not an expression as functions cannot be values in WDL.
* Fixed the CLI tools to not print colors when output is not to a terminal.
* Fixed incorrect parsing of some expressions which required the interior to be non-empty (e.g. `foo[]` is not a valid index expression).

Before submitting this PR, please make sure:

- [x] You have added a few sentences describing the PR here.
- [x] You have added yourself or the appropriate individual as the assignee.
- [x] You have added at least one relevant code reviewer to the PR.
- [x] Your code builds clean without any errors or warnings.
- [x] You have added tests (when appropriate).
- [x] You have updated the README or other documentation to account for these
      changes (when appropriate).
- [x] You have added an entry to the relevant `CHANGELOG.md` (see
      ["keep a changelog"] for more information).
- [x] Your commit messages follow the [conventional commit] style.

[conventional commit]: https://www.conventionalcommits.org/en/v1.0.0/#summary
["keep a changelog"]: https://keepachangelog.com/en/1.0.0/

### CODE REVIEW

**Comment by peterhuene** [comment]
Opened #164 to track the missing 1.2 features.

**Review by a-frantz — COMMENTED** [review]
2 Qs:
1. does merging this bring us up to parity with `miniwdl check`?
2. I don't see a test for a forward reference, but from my understanding of this code and the spec those should be parseable.
```
task foo {
    input {
        String arg = if bool then "--arg" else ""
        Boolean bool
    }
    command <<<>>>
}
```

File: wdl-analysis/src/task/v1.rs [hunk 1]
```diff
@@ -0,0 +1 @@
+//! Task graph representation for a V1 AST.
```
Comment (claymcleod) [inline wdl-analysis/src/task/v1.rs:1]: Not sure about this as well.

File: wdl-analysis/src/scope.rs [hunk 2]
```diff
@@ -221,28 +163,25 @@ pub struct Struct {
     /// This is either the name of a struct definition (local) or an import's
     /// URI or alias (imported).
     span: Span,
+    /// The offset of the CST node from the start of the document.
+    ///
+    /// This is used to adjust
```
Comment (claymcleod) [inline wdl-analysis/src/scope.rs:168]: Used to adjust?

File: wdl-analysis/src/stdlib.rs [hunk 3]
```diff
@@ -1581,6 +1610,16 @@ pub static STDLIB: LazyLock<StandardLibrary> = LazyLock::new(|| {
                             .parameter(PrimitiveTypeKind::String)
                             .ret(PrimitiveTypeKind::String)
                             .build(),
+                        // This overload isn't explicitly specified in the spec, but the spec
```
Comment (claymcleod) [inline wdl-analysis/src/stdlib.rs:1613]: Should we propose updates to the spec for these? Or no?

File: wdl-analysis/tests/analysis/conflicting-decl-names/source.diagnostics [hunk 4]
```diff
@@ -5,7 +5,7 @@ error: conflicting output name `x`
    │             - the input with the conflicting name is here
    ·
 13 │         String x = "x"
-   │                ^ this conflicts with a input of the same name
+   │                ^ this output conflicts with a previously used name
```
Comment (claymcleod) [inline wdl-analysis/tests/analysis/conflicting-decl-names/source.diagnostics:8]: this output name?

File: wdl-analysis/tests/analysis/name-reference-cycle/source.diagnostics [hunk 5]
```diff
@@ -0,0 +1,9 @@
+error: name reference cycle detected
+  ┌─ tests/analysis/name-reference-cycle/source.wdl:6:9
+  │
+6 │     Int a = b
+  │         ^ this name participates in the cycle
+7 │     Int b = c
```
Comment (claymcleod) [inline wdl-analysis/tests/analysis/name-reference-cycle/source.diagnostics:6]: Should this be flagged as participating in the cycle?

**Review by claymcleod — APPROVED** [review]
When those are addressed, looks great! Admittedly, there is a ton here, so I may not have caught everything—I sort of ran with the assumption that the approach you took is solid and we can continue to tweak it in the future if need be (mainly because I don't have any concrete experience with better ways to do it).

File: wdl-analysis/src/scope/v1.rs [hunk 6]
```diff
@@ -175,704 +201,1013 @@ fn call_conflict(name: &Ident, first: NameContext, suggest_fix: bool) -> Diagnos
...
+            DocumentItem::Import(_) | DocumentItem::Struct(_) => continue,
+            DocumentItem::Task(definition) => {
+                if let Some(task) = document.tasks.get(definition.name().as_str()) {
+                    type_check_task(document, &definition, task.scope, diagnostics);
                 }
             }
-            None => {
-                return Err(Some(import_missing_version(span)));
+            DocumentItem::Workflow(_) => {
+                // TODO: implement
```
Comment (claymcleod) [inline wdl-analysis/src/scope/v1.rs:924]: Just a note that this is still TODO as well.

**Comment by peterhuene** [comment]
@a-frantz:

> does merging this bring us up to parity with miniwdl check?

It does not yet; once I complete workflow type checking, it will.

> I don't see a test for a forward reference, but from my understanding of this code and the spec those should be parseable.

I added tests for cycle detection, but you're right, it'd be good to have one that passes without errors with a forward reference.  I'll get that added!

File: wdl-analysis/src/lib.rs [hunk 7]
```diff
@@ -10,14 +10,13 @@
 #![warn(rustdoc::broken_intra_doc_links)]
 
 mod analyzer;
+// pub mod evaluation;
```
[the remark this answers is not in the document]
Comment (peterhuene) [reply wdl-analysis/src/lib.rs:13]: It was not, good catch!

File: wdl-analysis/src/task/v1.rs (hunk 1 above)
Comment (peterhuene) [reply wdl-analysis/src/task/v1.rs:1]: This is also a dead code from very early on which was accidentally committed; I'll remove.

File: wdl-analysis/src/scope.rs (hunk 2 above)
Comment (peterhuene) [reply wdl-analysis/src/scope.rs:168]: Whoops, must have context switched away right in the middle of completing that sentence 😅. I'll fix!

File: wdl-analysis/src/stdlib.rs (hunk 3 above)
Comment (peterhuene) [reply wdl-analysis/src/stdlib.rs:1613]: It's _mentioned_ in the spec in the same section, but there probably should be explicit overloads for them in the list of signatures. I'll file a spec issue.

File: wdl-analysis/tests/analysis/name-reference-cycle/source.diagnostics (hunk 5 above)
Comment (peterhuene) [reply wdl-analysis/tests/analysis/name-reference-cycle/source.diagnostics:6]: The DFS stops at the first place it encounters a cyclic reference, so we only report on two locations: the starting point and the node that we've encountered which was previously visited.

If we want, we could walk _a_ path between the two nodes and report on each node we encounter.

File: wdl-analysis/src/scope/v1.rs (hunk 6 above)
Comment (peterhuene) [reply wdl-analysis/src/scope/v1.rs:924]: I'll file a tracking issue to complete static analysis by implementing workflow type checks.

File: wdl-analysis/src/stdlib.rs (hunk 3 above)
Comment (peterhuene) [reply wdl-analysis/src/stdlib.rs:1613]: Opened https://github.com/openwdl/wdl/issues/683.

File: wdl-analysis/src/scope/v1.rs (hunk 6 above)
Comment (peterhuene) [reply wdl-analysis/src/scope/v1.rs:924]: Opened https://github.com/stjude-rust-labs/wdl/issues/167.

**Review by a-frantz — APPROVED** [review]

**Pull request merged**